docs/proposals/scheduling/20230415-numa-topology-scheduling.md

```diff
@@ -0,0 +1,238 @@
+
+### Changes in the Scheduler component
+
+#### Add NUMATopology plugin
+
+Koodinator Scheduler needs to introduce a new plugin NUMATopology, which is responsible for coordinating the allocation of resources between different plugins and deciding on a suitable NUMA Node based on NUMA Topology Policy, and ensuring that the NUMA Node based on the decision can allocate resources needed by Pods. And by Score mechanism, select the most suitable node at the cluster level.
+
+NUMATopology will require plugins that perceive NUMA Topology to implement a new interface called NUMATopologyHintProvider. NUMATopologyHintProvider is borrowed from kubelet’s Topology Manager. Existing plugins NodeNUMAResource/DeviceShare need to implement this new interface.
+
+In the Filter phase, the NUMATopology plugin will call NUMATopologyHintProvider.GetPodTopologyHints of each HintProvider to obtain the NUMA Node Hint corresponding to each resource supported by this HintProvider, indicating which NUMA Node these resources can be allocated on. The NUMATopology plugin will integrate these Hints according to Policy and calculate a NUMA Node that can satisfy all resources. If there is no such NUMA Node, it means that this node cannot participate in subsequent scheduling allocation. If there is such a NUMA Node, it is called Preferred NUMA Node. Then continue to call NUMATopologyHintProvider.Allocate method to confirm whether resources can be allocated on Preferred NUMA Node. If not, it is considered that this node cannot participate in scheduling allocation either. If this node has a Preferred NUMA Node that meets the Pod’s requirements, it will be saved in CycleState for subsequent Score/Reserve phases.
+
+By default, MostAllocated algorithm is used for scoring, which prioritizes nodes with fewer remaining NUMA Nodes. It also supports scoring based on strategies specified by the node.koordinator.sh/numa-allocate-strategy label on nodes, but only supports two strategies corresponding to this label: MostAllocated and LeastAllocated.
+
+MostAllocated and LeastAllocated strategies are both calculated based on the relationship between the number of used NUMA Nodes and the total number of NUMA Nodes used. Taking MostAllocated as an example, if a node has 4 NUMA Nodes but one has been allocated completely and one more is needed during this scheduling process, then the calculation formula is as follows: *Score = (2 * 100)/4 = 50*
```

Comment:
Do we need to prioritize the absolute number of available NUMA Nodes? There may be machines with different NUMA nodes, and in some cases, the machine with the most numa nodes is preferred.

Reply:
Good idea. It can be used as a follow-up optimization item.

Reply:
OK